### ISSUE TITLE
[ENHANCEMENT] Shutdown grace period of 30s not valid for VMs that need a long time to shutdown

### ISSUE THREAD

**Issue by Martin-Weiss** [harvester/harvester#2466]
**Describe the bug**
It seems that the current shutdown wait time (before hard power of) for a guest VM is 30s.
It is not a valid approach to power of any VM during shutdown if it is not powered of after 30s.
(even some RKE2 nodes need much more time)

Every "hard" power off is a risk for data in the running machine - so we should avoid any "automated" hard power off as long and as good as we can. The "hard power off" should be the administrators decision.

**To Reproduce**
Deply a VM with a process that takes >30s to shutdown.
Shut down the VM.
Realize it is powered of before all processes could stop gracefully.

**Expected behavior**
The default should be to wait forever before the machine is "hard killed".

Beside the "shutdown graceful stop" button in the UI we also need a "forcefull power off" button in case the machine hangs or does not react on the ACPI based shutdown command.

**Environment**
 - Harvester ISO version: 1.0.2

**Comment by FrankYang0529** [comment]
Hi @Martin-Weiss, from [KubeVirt Virtual Machine Graceful Shutdown document](https://github.com/kubevirt/kubevirt/blob/main/docs/graceful-shutdown.md), it looks like we can't have a "shutdown graceful stop" to wait forever before the machine is "hard killed", but we can let users modify `terminationGracePeriodSeconds` to adjust a waiting period before "hard killed". We can have a global default setting and users can give a different value when they create a new VM.

For "forcefull power off" function, we can add an API to trigger stop and then force delete VM pod.

**Comment by Martin-Weiss** [comment]
Sounds valid. In general data protection should have highest priority and hard killing a machine will always be a risk for data and cause loss of grey data. 
So setting the terminationGracePeriodSeconds to something like i.e. 7 days might be a compromise while we also might need an additional button in the UI to force the power off in case this timer is waiting..
And yes - we need to make this configurable via global settings and per VM / per Namespace settings I assume.

**Comment by TachunLin** [comment]
Verified **fixed** on `master-b49e264e-head` (05/22). Close this issue.

### Result
#### [PASS] Test Case 1: 
If users don't set default-termination-grace-period-seconds setting, terminationGracePeriodSeconds in a VM is 30 seconds

* The `default-vm-termination-grace-period-seconds` settings is `120` seconds
  ![image](https://github.com/harvester/harvester/assets/29251855/afbc2771-47e8-4daf-aef6-03fe8d3b3d3a)

* And also the `Termination Grace Period` is `120` seconds on VM advance options 
  ![image](https://github.com/harvester/harvester/assets/29251855/8533a08c-ac75-43f9-83c0-75e88737df82)

* After vm created, we can see the `spec.template.spec.terminationGracePeriodSeconds` in the VM is `120`
  ![image](https://github.com/harvester/harvester/assets/29251855/126ab1f9-330d-45e4-8856-59d356593dca)

#### [PASS] Test Case 2
`default-termination-grace-period-seconds` setting can't be negative.

* When we set negative value to  the `default-vm-termination-grace-period-seconds`, it will prompt message to prevent to proceed
![image](https://github.com/harvester/harvester/assets/29251855/d8756f19-db8a-464a-ad3d-e7f09ccc2301)

* Also when we edit the value in `spec.template.spec.terminationGracePeriodSeconds` it will prompt message `value: can't be negative`
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # settings.harvesterhci.io "default-vm-termination-grace-period-seconds" was not valid:
  # * value: can't be negative
  #
  apiVersion: harvesterhci.io/v1beta1
  default: "120"
  kind: Setting
  metadata:
    creationTimestamp: "2023-05-22T07:14:38Z"
    generation: 8
    name: default-vm-termination-grace-period-seconds
    resourceVersion: "143138"
    uid: 896f2326-c7e5-4914-bf7e-e7b0544472e5
  status: {}
  value: "-1"
  ```

#### [PASS] Test Case 3
`default-termination-grace-period-seconds` setting should be able to be parsed as an integer.

* When we set string value to the `default-vm-termination-grace-period-seconds` and `spec.template.spec.terminationGracePeriodSeconds` it both prompt message to prevent to proceed
  ![image](https://github.com/harvester/harvester/assets/29251855/c26b78f2-335f-4397-a799-be710ed88cf0)


  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # settings.harvesterhci.io "default-vm-termination-grace-period-seconds" was not valid:
  # * value: strconv.ParseInt: parsing "test": invalid syntax
  #
  apiVersion: harvesterhci.io/v1beta1
  default: "120"
  kind: Setting
  metadata:
    creationTimestamp: "2023-05-22T07:14:38Z"
    generation: 8
    name: default-vm-termination-grace-period-seconds
    resourceVersion: "143138"
    uid: 896f2326-c7e5-4914-bf7e-e7b0544472e5
  status: {}
  value: "test"
  ```

#### [PASS] Test Case 4
Set default-termination-grace-period-seconds setting and terminationGracePeriodSecondsin a new VM should follow the setting

* When we set the `default-vm-termination-grace-period-seconds` set to `120` and create a VM. After the VM running, the `spec.template.spec.terminationGracePeriodSeconds` set to `120` accordingly and save to the yaml 

  ![image](https://github.com/harvester/harvester/assets/29251855/ef221f3a-c6df-4890-8a1a-0a3a4e1d114d)

#### [PASS] Test Case 5
The controller should wait for terminationGracePeriodSeconds, before really removing the VM

* With set the `default-vm-termination-grace-period-seconds` set to `120` when we create a running process and stop the VM. We can see OS will wait for 110 seconds. 
  ![image](https://github.com/harvester/harvester/assets/29251855/34cf2a65-4aaa-4c43-a385-ab739210f6cf)
* And controller will wait for 120 seconds, so the VM can graceful shutdown.
  ![image](https://github.com/harvester/harvester/assets/29251855/ff697fd0-062d-4f82-8922-dedd8574085d)

#### [PASS] Test Case 6
Users can force stop the VM if they don't wait for a graceful shutdown.

* When we click the `force stop` in the view API of the VM in the `Stopping` state
  ![image](https://github.com/harvester/harvester/assets/29251855/ef7ad189-f99e-4ee7-9126-9830d0df9bb0)

* The VM will immediately shutdown
  ![image](https://github.com/harvester/harvester/assets/29251855/08e45d8c-8763-453a-babc-2c7007712c4c)

#### [PASS] Test Case 7
We **can't** set `empty` value to `default-vm-termination-grace-period-seconds`

* When we try to set `empty` value to `default-vm-termination-grace-period-seconds`, it will prompt message to prohibit.
  ![image](https://github.com/harvester/harvester/assets/29251855/0c8f576d-48d2-4d78-bed8-1fccecc3a6b0)

#### [PASS] Test Case 8
We can set `0` integer value to `default-vm-termination-grace-period-seconds`

* We can set `0` to `default-vm-termination-grace-period-seconds` and apply to vm `terminationGracePeriodSeconds` accordingly.
  ![image](https://github.com/harvester/harvester/assets/29251855/dc755788-8d02-436f-9fc4-7f296d8c51b2)

### Test Information
* Test Environment: single nodes Harvester on local kvm machine
* Harvester version: `master-b49e264e-head` (05/22)

### Verify Steps
https://github.com/harvester/tests/issues/732#issue-1604415795
